Change AbstractBuild to Run

rodrigc opened the pull request:
This is for compatibility with workflow/pipeline plugin.

rodrigc commented:
This patch was confirmed by E K to work with the pipeline (formerly workflow) plugin:

https://groups.google.com/d/msg/jenkinsci-users/gwhMuq4KPi8/NKVh_dPQFwAJ

dnozay commented:
@rodrigc can you please add a test? Looks like there is a repro testcase on the forum. Thank you

rodrigc commented:
The test involves invoking a build-flow from inside pipeline (formerly workflow) DSL.
Is it OK to add a test-only dependency on the pipeline plugin?

dnozay commented:
@rodrigc yes, please add a comment inside the added block in the pom. thanks you.

rodrigc commented:
@dnozay I have added the test which uses pipeline (formerly workflow).  In addition, I fixed some of the tests, which were failing.

dnozay commented on src/test/groovy/com/cloudbees/plugins/flow/ConcurrencyTest.groovy:
```diff
@@ -97,20 +102,21 @@ class ConcurrencyTest extends DSLTestCase {
+        flow.onCreatedFromScratch() // need this to updateTransientActions 
 
-        def sfr1 = flow.scheduleBuild2(0)
+        def sfr1 = flow.scheduleBuild2(0, new Cause.UserIdCause())
```
hmmmm, well, the first argument is supposed to be the `quietPeriod`, I'm not sure it should be `0` based on the test description.

dnozay commented:
The only thing blocking this to be merged is the concurrency test that is now bogus and needs to get fixed somehow.

rodrigc commented:
@dnozay I think the concurrency test which is failing should just be deleted.  See: https://github.com/jenkinsci/git-plugin/commit/4fe33a1fa31459563b7d9e4c4c652b20bde6df4d

rodrigc commented:
@dnozay can you take a look at this pull request?  I believe that I have addressed all your concerns.
This change will allow people to call a build-flow job from within a pipeline.  It is a way to allow people to slowly migrate to pipeline.

dnozay commented:
This looks fine to me. Please feel free to take maintainership and perform a new release.

dnozay commented:
@rodrigc 

Since I am currently looking for someone else to take over as a maintainer; I will not be the one merging this change. Please check <http://jenkins-ci.org/pull-request-greeting>.

Side note: I think introducing new features in this plugin is risky and takes away resources that could be better spent on the [pipeline plugin](https://github.com/jenkinsci/pipeline-plugin). However, please feel free to merge if you are interested in carrying on the flag.
--
Well, since this adds more compatibility to the workflow / pipeline plugin, that change would certainly help the transition. Thank you for removing the flaky test.

rodrigc commented:
I agree with you that new features should not be added to buildflow-plugin.
People should start transitioning to Pipeline, because that is where the core Jenkins developers
are focusing their efforts.
Bugfixes (which do not add new features) to build-flow plugin are probably OK, since people
have created many build flows in the past few years when workflow/pipeline did not exist.

At least my change will facilitate transition to Pipeline and does not add new features.

rodrigc commented:
@dnozay If you are OK with me taking maintainership of this plugin, then I will do it.  If you agree, please respond on this thread, and I will point the Jenkins admins to it.
I will consider this plugin as deprecated in favor of Pipeline, but apply minimal bugfixes, since
many people have written build-flows over the years when Pipeline did not exist.

dnozay commented:
@rodrigc go for it. I have already updated the Readme, wiki, and posted on the ML. Thank you for stepping up. 👍
